From 6a0fcb338c61beacf68feae7d3afd781e3585e02 Mon Sep 17 00:00:00 2001
From: Bing Xue <bingxue@google.com>
Date: Tue, 11 Dec 2018 10:32:33 -0800
Subject: [PATCH] tensorflow: Allow dlopen for libneuralnetworks.so

TFLite does not support libneuralnetworks outside of android builds.
This CL allows dlopen for libneuralnetworks.so for all builds and preserves the
default behavior to fail silently when libneuralnetworks is not found.

The corresponding patch in tensorflow upstream is at:
https://github.com/tensorflow/tensorflow/commit/17b71f26abaf3d96d29f346055333096a5f8d03d#diff-87f1e95e86662c048aae7947ba70f653

crbug.com/924331 tracks this effort on our side.
---
 tensorflow/contrib/lite/nnapi/NeuralNetworksShim.h | 6 ++----
 tensorflow/contrib/lite/nnapi_delegate.cc          | 2 ++
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/tensorflow/contrib/lite/nnapi/NeuralNetworksShim.h b/tensorflow/contrib/lite/nnapi/NeuralNetworksShim.h
index becd1f615f..5e667e7baa 100644
--- a/tensorflow/contrib/lite/nnapi/NeuralNetworksShim.h
+++ b/tensorflow/contrib/lite/nnapi/NeuralNetworksShim.h
@@ -35,12 +35,10 @@ inline void* loadLibrary(const char* name) {
   // TODO: change RTLD_LOCAL? Assumes there can be multiple instances of nn
   // api RT
   void* handle = nullptr;
-#ifdef __ANDROID__
   handle = dlopen(name, RTLD_LAZY | RTLD_LOCAL);
-  if (handle == nullptr) {
-    NNAPI_LOG("nnapi error: unable to open library %s", name);
+  if (handle != nullptr) {
+    NNAPI_LOG("nnapi: success open library %s", name);
   }
-#endif
   return handle;
 }
 
diff --git a/tensorflow/contrib/lite/nnapi_delegate.cc b/tensorflow/contrib/lite/nnapi_delegate.cc
index fad08bbfe6..7251378e13 100644
--- a/tensorflow/contrib/lite/nnapi_delegate.cc
+++ b/tensorflow/contrib/lite/nnapi_delegate.cc
@@ -501,9 +501,11 @@ void AddOpsAndParams(tflite::Interpreter* interpreter,
         break;
     }
 
+#ifdef __ANDROID__
     if (nnapi_version == 11 && kAndroidSdkVersion < 28) {
       FATAL("Op %d needs NNAPI1.1", builtin);
     }
+#endif // __ANDROID__
 
     // Add the operation.
     CHECK_NN(ANeuralNetworksModel_addOperation(
-- 
2.20.1.791.gb4d0f1c61a-goog

